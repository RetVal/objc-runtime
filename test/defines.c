/* 
TEST_CONFIG MEM=mrc LANGUAGE=c

TEST_BUILD
$DIR/defines.sh '$C{TESTINCLUDEDIR}' '$C{TESTLOCALINCLUDEDIR}' '$C{COMPILE_C}' '$C{COMPILE_CXX}' '$C{COMPILE_M}' '$C{COMPILE_MM}' '$VERBOSE'
$C{COMPILE_C} $DIR/defines.c -o defines.exe
END

TEST_BUILD_OUTPUT
(.|\n)*No unexpected #defines found\.
END
 */


#include "test.h"

int main()
{
    succeed(__FILE__);
}
