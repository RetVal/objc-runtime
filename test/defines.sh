#!/bin/sh

# Check ObjC headers for unwanted defines exposed to clients.

TESTINCLUDEDIR=$1; shift
TESTLOCALINCLUDEDIR=$1; shift
COMPILE_C=$1; shift
COMPILE_CXX=$1; shift
COMPILE_M=$1; shift
COMPILE_MM=$1; shift
VERBOSE=$1; shift

# stop after any command error
set -e

# echo commands when verbose
if [ "$VERBOSE" != "0" ]; then
    set -x
fi

FILES="$TESTINCLUDEDIR/objc/*.h $TESTLOCALINCLUDEDIR/objc/*.h"
CFLAGS='-fsyntax-only -Wno-unused-function -D_OBJC_PRIVATE_H_'

INCLUDES=$(grep -h '#include' $FILES | grep -v '<objc/' \
    | sed 's/^#include \(.*\)$/#if __has_include(\1)\n#include \1\n#endif/g')

sort $(dirname $0)/defines.expected > defines.expected

ERROR=

extract_defines() {
    echo "$INCLUDES" | $1 - -dM -E \
        | sed 's/\(#define [_A-Za-z][_A-Za-z0-9]*\).*/\1/g' \
        | sort | uniq
}

get_new_lines() {
    diff -u -U 0 $1 $2 | grep -v "^+++" | grep "^+" | cut -c2- || true
}

run_test() {
    extract_defines "$1 $CFLAGS" > base-defines
    extract_defines "$1 $CFLAGS $FILES" > objc-defines
    get_new_lines base-defines objc-defines > objc-defines-only
    if [[ ! -s objc-defines-only ]]; then
        echo "ERROR: objc-defines-only is somehow empty."
        exit 1
    fi
    get_new_lines defines.expected objc-defines-only > objc-defines-unexpected
    if [[ -s objc-defines-unexpected ]]; then
        echo "ERROR: unknown #defines found in headers. If these are expected, add them to test/defines.expected."
        echo "$1"
        cat objc-defines-unexpected
        ERROR=1
    fi
    rm base-defines objc-defines objc-defines-only objc-defines-unexpected
}

run_test "$COMPILE_C $CFLAGS"
run_test "$COMPILE_CXX $CFLAGS"
run_test "$COMPILE_M $CFLAGS"
run_test "$COMPILE_MM $CFLAGS"
for STDC in '99' '11' ; do
    run_test "$COMPILE_C $CFLAGS -std=c$STDC"
    run_test "$COMPILE_M $CFLAGS -std=c$STDC"
done
for STDCXX in '98' '03' '11' '14' '17' ; do
    run_test "$COMPILE_CXX $CFLAGS -std=c++$STDCXX"
    run_test "$COMPILE_MM $CFLAGS -std=c++$STDCXX"
done

if [[ $ERROR == "" ]]; then
    echo "No unexpected #defines found."
else
    echo "Unknown #defines found in headers."
fi
