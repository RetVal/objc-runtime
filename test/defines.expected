#define ARITH_SHIFT
#define AUTORELEASEPOOL_VERSION
#define CHECK_AUTORELEASEPOOL
#define CONFIG_USE_PREOPT_CACHES
#define ISSELECTOR
#define IV
#define NAMEOF
#define NO
#define NSINTEGER_DEFINED
#define NSIntegerMax
#define NSIntegerMin
#define NSUIntegerMax
#define NS_DESIGNATED_INITIALIZER
#define NS_ENFORCE_NSOBJECT_DESIGNATED_INITIALIZER
#define NX_MAPNOTAKEY
#define Nil
#define OBJC_ADDLOADIMAGEFUNC_DEFINED
#define OBJC_ADDLOADIMAGEFUNC2_DEFINED
#define OBJC_API_VERSION
#define OBJC_ARC_UNAVAILABLE
#define OBJC_ARM64_UNAVAILABLE
#define OBJC_ASSERT
#define OBJC_AVAILABLE
#define OBJC_BOOL_DEFINED
#define OBJC_BOOL_IS_BOOL
#define OBJC_BOOL_IS_CHAR
#define OBJC_COLD
#define OBJC_DECLARED_ABSOLUTE_PACKED_ISA_CLASS_MASK
#define OBJC_DEPRECATED
#define OBJC_DYNAMIC_CLASSES
#define OBJC_ENUM
#define OBJC_EXIT_REASON_CLASS_RO_SIGNING_REQUIRED
#define OBJC_EXIT_REASON_GC_NOT_SUPPORTED
#define OBJC_EXIT_REASON_UNSPECIFIED
#define OBJC_EXPORT
#define OBJC_EXTERN
#define OBJC_GC_DEPRECATED
#define OBJC_GC_UNAVAILABLE
#define OBJC_GETCLASSHOOK_DEFINED
#define OBJC_HASH_AVAILABILITY
#define OBJC_HAS_DUPLICATE_CLASS
#define OBJC_HAVE_INDEXED_NONPOINTER_ISA
#define OBJC_HAVE_NONPOINTER_ISA
#define OBJC_HAVE_PACKED_NONPOINTER_ISA
#define OBJC_HAVE_TAGGED_POINTERS
#define OBJC_IMPORT
#define OBJC_INLINE
#define OBJC_ISA_AVAILABILITY
#define OBJC_ISUNIQUELYREFERENCED_DEFINED
#define OBJC_MAP_AVAILABILITY
#define OBJC_MAX_CLASS_SIZE
#define OBJC_MSB_TAGGED_POINTERS
#define OBJC_NOESCAPE
#define OBJC_NORETURN
#define OBJC_NO_GC
#define OBJC_NO_GC_API
#define OBJC_OLD_DISPATCH_PROTOTYPES
#define OBJC_OPTIONS
#define OBJC_OSX_AVAILABLE_OTHERS_UNAVAILABLE
#define OBJC_OSX_DEPRECATED_OTHERS_UNAVAILABLE
#define OBJC_REALIZECLASSFROMSWIFT_DEFINED
#define OBJC_REFINED_FOR_SWIFT
#define OBJC_RETURNS_RETAINED
#define OBJC_ROOT_CLASS
#define OBJC_SETCUSTOMDEALLOCINITIATION_DEFINED
#define OBJC_SETHOOK_LAZYCLASSNAMER_DEFINED
#define OBJC_SPLIT_TAGGED_POINTERS
#define OBJC_SUPER
#define OBJC_SWIFT_UNAVAILABLE
#define OBJC_UNAVAILABLE
#define OBJC_VISIBLE
#define OBJC_WEAK_FORMATION_CALLOUT_DEFINED
#define OBJC_WILLINITIALIZECLASSFUNC_DEFINED
#define PROTECT_AUTORELEASEPOOL
#define SELNAME
#define SELUID
#define YES
#define _C_ARY_B
#define _C_ARY_E
#define _C_ATOM
#define _C_ATOMIC
#define _C_BFLD
#define _C_BOOL
#define _C_BYCOPY
#define _C_BYREF
#define _C_CHARPTR
#define _C_CHR
#define _C_CLASS
#define _C_COMPLEX
#define _C_CONST
#define _C_DBL
#define _C_FLT
#define _C_GNUREGISTER
#define _C_ID
#define _C_IN
#define _C_INOUT
#define _C_INT
#define _C_INT128
#define _C_LNG
#define _C_LNG_DBL
#define _C_LNG_LNG
#define _C_ONEWAY
#define _C_OUT
#define _C_PTR
#define _C_SEL
#define _C_SHT
#define _C_STRUCT_B
#define _C_STRUCT_E
#define _C_UCHR
#define _C_UINT
#define _C_UINT128
#define _C_ULNG
#define _C_ULNG_LNG
#define _C_UNDEF
#define _C_UNION_B
#define _C_UNION_E
#define _C_USHT
#define _C_VECTOR
#define _C_VOID
#define _NSOBJECT_INTERNAL_H
#define _OBJC_ABI_H
#define _OBJC_ASSOCIATION_SYSTEM_OBJECT
#define _OBJC_AUTO_H_
#define _OBJC_GDB_H
#define _OBJC_INTERNAL_H
#define _OBJC_LIST_H_
#define _OBJC_LITTLE_HASHTABLE_H_
#define _OBJC_LOAD_H_
#define _OBJC_MAPTABLE_H_
#define _OBJC_MESSAGE_H
#define _OBJC_NSOBJCRUNTIME_H_
#define _OBJC_NSOBJECT_H_
#define _OBJC_OBJC_API_H_
#define _OBJC_OBJC_H_
#define _OBJC_OBJECT_H_
#define _OBJC_PROTOCOL_H_
#define _OBJC_RUNTIME_H
#define _OBJC_SUPPORTED_INLINE_REFCNT
#define _OBJC_SUPPORTED_INLINE_REFCNT_LOGIC
#define _OBJC_SUPPORTED_INLINE_REFCNT_LOGIC_BLOCK
#define _OBJC_SUPPORTED_INLINE_REFCNT_WITH_DEALLOC2MAIN
#define _OBJC_TAG_CONSTANT_POINTER_MASK
#define _OBJC_TAG_EXT_INDEX_MASK
#define _OBJC_TAG_EXT_INDEX_SHIFT
#define _OBJC_TAG_EXT_MASK
#define _OBJC_TAG_EXT_PAYLOAD_LSHIFT
#define _OBJC_TAG_EXT_PAYLOAD_RSHIFT
#define _OBJC_TAG_EXT_SLOT_COUNT
#define _OBJC_TAG_EXT_SLOT_MASK
#define _OBJC_TAG_EXT_SLOT_SHIFT
#define _OBJC_TAG_INDEX_MASK
#define _OBJC_TAG_INDEX_SHIFT
#define _OBJC_TAG_MASK
#define _OBJC_TAG_NO_OBFUSCATION_MASK
#define _OBJC_TAG_PAYLOAD_LSHIFT
#define _OBJC_TAG_PAYLOAD_RSHIFT
#define _OBJC_TAG_SLOT_COUNT
#define _OBJC_TAG_SLOT_MASK
#define _OBJC_TAG_SLOT_SHIFT
#define _OBJC_TRAMPOLINES_H
#define __BRIDGEOS_AVAILABLE
#define __BRIDGEOS_DEPRECATED
#define __BRIDGEOS_UNAVAILABLE
#define __OBJC_EXCEPTION_H_
#define __OBJC_SNYC_H_
#define __autoreleasing
#define __DARWIN_NULL
#define nil
